### PR TITLE
Msf::Post::Windows: Replace load_extapi with ExtAPI capability check

### PR DESCRIPTION
**Pull request by bcoles**
See https://github.com/rapid7/metasploit-framework/pull/16921#issuecomment-1227746612 for context.

If ExtAPI can be loaded by Meterpreter then ExtAPI should already be loaded automatically.

Replacing `load_extapi` with a more granular capability check approach is cleaner and more intuitive.

Tested LDAP using `post/windows/gather/enum_ad_computers`.

Tested WMIC using `post/windows/gather/wmic_command`.

~~The `Msf::Post::Windows::Service` library still uses `load_extapi`; however, this will be removed by #16928.~~ Merged.

Several modules still use `load_extapi` which will need to be reviewed (in a separate PR) before the `load_extapi` method can be removed from Framework:

```
# grep -rn load_extapi modules/
modules/exploits/windows/local/wmi.rb:75:    if load_extapi
modules/exploits/windows/local/wmi.rb:89:      if load_extapi
modules/exploits/windows/local/wmi.rb:134:      unless load_extapi
modules/post/windows/gather/credentials/domain_hashdump.rb:155:    load_extapi
modules/post/windows/gather/enum_patches.rb:41:    unless load_extapi
```

~~`load_extapi` pending removal from `modules/post/windows/gather/enum_patches.rb` in #17003.~~ Merged.

### CODE REVIEW

File: lib/msf/core/post/windows/wmic.rb [hunk 1]
```diff
@@ -47,7 +45,7 @@ def wmic_query(query, server=datastore['RHOST'])
       end
     end
 
-    if extapi && !is_system?
+    if !is_system? && session.commands.include?(Rex::Post::Meterpreter::Extensions::Extapi::COMMAND_ID_EXTAPI_CLIPBOARD_SET_DATA)
```
Comment (dwelch-r7) [inline lib/msf/core/post/windows/wmic.rb:48]:
```suggestion
    if session.commands.include?(Rex::Post::Meterpreter::Extensions::Extapi::COMMAND_ID_EXTAPI_CLIPBOARD_SET_DATA) && !is_system? 
```
Think we should probably keep the order the same here and for the other one below

Comment (bcoles) [reply lib/msf/core/post/windows/wmic.rb:48]: y

Comment (bcoles) [reply lib/msf/core/post/windows/wmic.rb:48]: changed

**Review by gwillcox-r7 — COMMENTED** [review]
I think the changes look good but may repeat the issue I mentioned at https://github.com/rapid7/metasploit-framework/pull/17003#discussion_r972461449 r.e normal sessions not being able to find these Rex::Post command ID definitions.

**Comment by gwillcox-r7** [comment]
With updates this now errors a lot cleaner on shell sessions:

```
msf6 payload(windows/x64/shell/bind_tcp) > use post/windows/gather/enum_ad_computers 
msf6 post(windows/gather/enum_ad_computers) > show otpions
[-] Invalid parameter "otpions", use "show -h" for more information
msf6 post(windows/gather/enum_ad_computers) > show options

Module options (post/windows/gather/enum_ad_computers):

   Name        Current Setting                         Required  Description
   ----        ---------------                         --------  -----------
   DOMAIN                                              no        The domain to query or distinguished name (e.g. DC=test,DC=com)
   FIELDS      dNSHostName,distinguishedName,descript  yes       FIELDS to retrieve.
               ion,operatingSystem,operatingSystemSer
               vicePack
   FILTER      (&(objectCategory=computer)(operatingS  yes       Search filter.
               ystem=*server*))
   MAX_SEARCH  500                                     yes       Maximum values to retrieve, 0 for all.
   SESSION                                             yes       The session to run this module on
   STORE_DB    false                                   yes       Store file in DB (performance hit resolving IPs).
   STORE_LOOT  false                                   yes       Store file in loot.

msf6 post(windows/gather/enum_ad_computers) > set DOMAIN DC=test,DC=com
DOMAIN => DC=test,DC=com
msf6 post(windows/gather/enum_ad_computers) > set DOMAIN DC=daforest,DC=local
DOMAIN => DC=daforest,DC=local
msf6 post(windows/gather/enum_ad_computers) > set SESSION 2
SESSION => 2
msf6 post(windows/gather/enum_ad_computers) > run

[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: shell
[-] Can't load wldap32: Session doesn't support Railgun!
[*] Post module execution completed
msf6 post(windows/gather/enum_ad_computers) > 
```

And with some more adjustments the wmic_command module also is being handled well with shell sessions:

```
msf6 post(windows/gather/wmic_command) > sessions

Active sessions
===============

  Id  Name  Type               Information                                      Connection
  --  ----  ----               -----------                                      ----------
  2         shell x64/windows  Shell Banner: Microsoft Windows [Version 10.0.2  172.28.94.235:46155 -> 172.28.82.203:4444 (172.2
                               0348.946] -----                                  8.82.203)

msf6 post(windows/gather/wmic_command) > show options

Module options (post/windows/gather/wmic_command):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   COMMAND                     no        WMIC command options.
   RESOURCE                    no        Full path to resource file to read commands from.
   RHOST      localhost        yes       Target address range
   SESSION    2                yes       The session to run this module on
   SMBDomain                   no        The Windows domain to use for authentication
   SMBPass                     no        The password for the specified username
   SMBUser                     no        The username to authenticate as
   TIMEOUT    10               yes       Timeout for WMI command in seconds

msf6 post(windows/gather/wmic_command) > run

[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: shell
[-] Post aborted due to failure: bad-config: The session this module is running against doesn't support the sysinfo command!
[*] Post module execution completed
msf6 post(windows/gather/wmic_command) > 
```

**Comment by gwillcox-r7** [comment]
Powershell shells also working well: 

```
msf6 payload(windows/powershell_bind_tcp) > to_handler
[*] Payload Handler Started as Job 2

[*] Started bind TCP handler against 172.28.82.203:4444
msf6 payload(windows/powershell_bind_tcp) > [*] Powershell session session 3 opened (172.28.94.235:46571 -> 172.28.82.203:4444) at 2022-09-20 12:59:46 -0500

msf6 payload(windows/powershell_bind_tcp) > previous
msf6 post(windows/gather/wmic_command) > sessions

Active sessions
===============

  Id  Name  Type                Information                                      Connection
  --  ----  ----                -----------                                      ----------
  2         shell x64/windows   Shell Banner: Microsoft Windows [Version 10.0.2  172.28.94.235:46155 -> 172.28.82.203:4444 (172.
                                0348.946] -----                                  28.82.203)
  3         powershell windows  Administrator @ WIN-BR0CCBA815B                  172.28.94.235:46571 -> 172.28.82.203:4444 (172.
                                                                                 28.82.203)

msf6 post(windows/gather/wmic_command) > set SESSION 3
SESSION => 3
msf6 post(windows/gather/wmic_command) > run

[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: powershell
[-] Post aborted due to failure: bad-config: The session this module is running against doesn't support the sysinfo command!
[*] Post module execution completed
msf6 post(windows/gather/wmic_command) > use post/windows/gather/enum_ad_computers 
msf6 post(windows/gather/enum_ad_computers) > set SESSION 3
SESSION => 3
msf6 post(windows/gather/enum_ad_computers) > run

[!] SESSION may not be compatible with this module:
[!]  * incompatible session type: powershell
[-] Can't load wldap32: Session doesn't support Railgun!
[*] Post module execution completed
msf6 post(windows/gather/enum_ad_computers) > 
```

**Comment by gwillcox-r7** [comment]
With Meterpreter sessions seems to work fine. Will upload these changes that I've been testing against locally now then land if tests pass: 

```
msf6 payload(windows/x64/meterpreter/bind_tcp) > to_handler
[*] Payload Handler Started as Job 7
msf6 payload(windows/x64/meterpreter/bind_tcp) > 
[*] Started bind TCP handler against 172.28.82.203:4444
[*] Sending stage (200774 bytes) to 172.28.82.203
[*] Meterpreter session 8 opened (172.28.94.235:32963 -> 172.28.82.203:4444) at 2022-09-20 13:05:27 -0500

msf6 payload(windows/x64/meterpreter/bind_tcp) > previous
msf6 post(windows/gather/enum_ad_computers) > to_handler
[-] Unknown command: to_handler
msf6 post(windows/gather/enum_ad_computers) > set SESSION 8 
SESSION => 8
msf6 post(windows/gather/enum_ad_computers) > show options

Module options (post/windows/gather/enum_ad_computers):

   Name        Current Setting                         Required  Description
   ----        ---------------                         --------  -----------
   DOMAIN      DC=daforest,DC=local                    no        The domain to query or distinguished name (e.g. DC=test,DC=com)
   FIELDS      dNSHostName,distinguishedName,descript  yes       FIELDS to retrieve.
               ion,operatingSystem,operatingSystemSer
               vicePack
   FILTER      (&(objectCategory=computer)(operatingS  yes       Search filter.
               ystem=*server*))
   MAX_SEARCH  500                                     yes       Maximum values to retrieve, 0 for all.
   SESSION     8                                       yes       The session to run this module on
   STORE_DB    false                                   yes       Store file in DB (performance hit resolving IPs).
   STORE_LOOT  false                                   yes       Store file in loot.

msf6 post(windows/gather/enum_ad_computers) > set DOMAIN DC=daforest,DC=com
DOMAIN => DC=daforest,DC=com
msf6 post(windows/gather/enum_ad_computers) > reload
[*] Reloading module...
msf6 post(windows/gather/enum_ad_computers) > run

Domain Computers
================

 dNSHostName                 distinguishedName             description  operatingSystem               operatingSystemServicePack
 -----------                 -----------------             -----------  ---------------               --------------------------
 WIN-BR0CCBA815B.daforest.c  CN=WIN-BR0CCBA815B,OU=Domain               Windows Server 2022 Standard
 om                           Controllers,DC=daforest,DC=
                             com

[*] Post module execution completed
msf6 post(windows/gather/enum_ad_computers) > set DOMAIN DC=daforest,DC=d
DOMAIN => DC=daforest,DC=d
msf6 post(windows/gather/enum_ad_computers) > run

[-] extapi_adsi_domain_query: Operation failed: 2147950635
[*] Post module execution completed
msf6 post(windows/gather/enum_ad_computers) > use post/windows/gather/wmic_command 
msf6 post(windows/gather/wmic_command) > show options

Module options (post/windows/gather/wmic_command):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   COMMAND                     no        WMIC command options.
   RESOURCE                    no        Full path to resource file to read commands from.
   RHOST      localhost        yes       Target address range
   SESSION    3                yes       The session to run this module on
   SMBDomain                   no        The Windows domain to use for authentication
   SMBPass                     no        The password for the specified username
   SMBUser                     no        The username to authenticate as
   TIMEOUT    10               yes       Timeout for WMI command in seconds

msf6 post(windows/gather/wmic_command) > set SESSION 8 
SESSION => 8
msf6 post(windows/gather/wmic_command) > run

[*] Running module against WIN-BR0CCBA815B
[*] Post module execution completed
msf6 post(windows/gather/wmic_command) > set COMMAND bios get serialnumber
COMMAND => bios get serialnumber
msf6 post(windows/gather/wmic_command) > run

[*] Running module against WIN-BR0CCBA815B
[*] Command output saved to: /home/gwillcox/.msf4/loot/20220920130759_default_172.28.82.203_host.command.wmi_318444.txt
[*] Post module execution completed
msf6 post(windows/gather/wmic_command) > cat /home/gwillcox/.msf4/loot/20220920130759_default_172.28.82.203_host.command.wmi_318444.txt
[*] exec: cat /home/gwillcox/.msf4/loot/20220920130759_default_172.28.82.203_host.command.wmi_318444.txt

SerialNumber                      
0272-6377-7766-0813-1404-7669-58  

msf6 post(windows/gather/wmic_command) > 
```

**Comment by gwillcox-r7** [comment]
# Release Notes
The `lib/msf/core/post/windows/accounts.rb`,  `lib/msf/core/post/windows/ldap.rb`, and `lib/msf/core/post/windows/wmic.rb` libraries have been updated to replace calls to `load_extapi` with ExtAPI compatibility checks which will check if the session supports ExtAPI, since if the sessions supports ExtAPI, it should already be loaded.

**Comment by bcoles** [comment]
> I think the changes look good but may repeat the issue I mentioned at [#17003 (comment)](https://github.com/rapid7/metasploit-framework/pull/17003#discussion_r972461449) r.e normal sessions not being able to find these Rex::Post command ID definitions.

Given that the ExtAPI mixin effectively does nothing (it contains only a `load_extapi` method which can eventually be removed) it may make more sense to add `require 'rex/post/meterpreter/extensions/extapi/command_ids'` to this mixin. That way modules and libraries which need the command IDs can simply import the mixin instead - which they already are in most cases.

**Pull request merged**